### PR TITLE
BZ2008990: Updates to ASH for internal load balancer

### PR DESCRIPTION
**Pull request by mjpytlak**
CP to 4.9

https://bugzilla.redhat.com/show_bug.cgi?id=2008990

Updates to Installing a cluster on Azure Stack Hub using ARM templates:

- Updated the Network load balancer row in the table that is under Azure Stack Hub account limits.
  - Number of components required by default value is now `3`.
  - Description now references `internal` and `external` load balancers.
 https://deploy-preview-36950--osdocs.netlify.app/openshift-enterprise/latest/installing/installing_azure_stack_hub/installing-azure-stack-hub-user-infra.html#installation-azure-limits_installing-azure-stack-hub-user-infra

- Updated Creating networking and load balancing components in Azure Stack Hub.
  - Intro text lists the `api` and `api-int` DNS records being required for load balancing.
  - Added step 3b to export the private IP for the internal load balancer
  - Updated step 3d and supporting terminal examples to specify that the DNS record is for the private zone.
  https://deploy-preview-36950--osdocs.netlify.app/openshift-enterprise/latest/installing/installing_azure_stack_hub/installing-azure-stack-hub-user-infra.html#installation-creating-azure-dns_installing-azure-stack-hub-user-infra

### CODE REVIEW

**Comment by mjpytlak** [comment]
@patrickdillon Ready for SME review. Please take a look when you have a moment. Thank you.

@codyhoag Please take a look at the changes I made to installation-azure-limits.adoc. Best I can tell, this module is only shared between the AZ and ASH topics. Per the BZ, Patrick requested these updates. What has me curious if the `ifndef` at line 148. Was it to preserve the original table row in case the module was used later for other platforms?

File: modules/installation-creating-azure-dns.adoc [hunk 1]
```diff
@@ -65,20 +69,34 @@ $ az deployment group create -g ${RESOURCE_GROUP} \
...
+ifdef::ash[]
+. Create an `api` DNS record and an `api-int` DNS record. When creating the API DNS records, the `${BASE_DOMAIN_RESOURCE_GROUP}` variable must point to the resource group where the DNS zone exists.
 endif::ash[]
 
 .. Export the following variable:
 +
 [source,terminal]
 ----
-$ export PUBLIC_IP=`az network public-ip list -g ${RESOURCE_GROUP} --query "[?name=='${INFRA_ID}-master-pip'] | [0].ipAddress" -o tsv`
+$ export PUBLIC_IP=$(az network public-ip list -g ${RESOURCE_GROUP} --query "[?name=='${INFRA_ID}-master-pip'] | [0].ipAddress" -o tsv)
```
Comment (codyhoag) [inline modules/installation-creating-azure-dns.adoc:88]:
```suggestion
$ export PUBLIC_IP=`az network public-ip list -g ${RESOURCE_GROUP} --query "[?name=='${INFRA_ID}-master-pip'] | [0].ipAddress" -o tsv`
```
For consistency with the other variable export commands.

File: modules/installation-creating-azure-dns.adoc [hunk 2]
```diff
@@ -65,20 +69,34 @@ $ az deployment group create -g ${RESOURCE_GROUP} \
...
 ----
+ifdef::ash[]
+.. Export the following variable:
++
+[source,terminal]
+----
+$ export PRIVATE_IP=$(az network lb frontend-ip show -g "$RESOURCE_GROUP" --lb-name "${INFRA_ID}-internal" -n internal-lb-ip --query "privateIpAddress" -o tsv)
```
Comment (codyhoag) [inline modules/installation-creating-azure-dns.adoc:95]:
```suggestion
$ export PRIVATE_IP=`az network lb frontend-ip show -g "${RESOURCE_GROUP}" --lb-name "${INFRA_ID}-internal" -n internal-lb-ip --query "privateIpAddress" -o tsv`
```
For consistency

File: modules/installation-creating-azure-dns.adoc [hunk 3]
```diff
@@ -65,20 +69,34 @@ $ az deployment group create -g ${RESOURCE_GROUP} \
...
+----
+endif::ash[]
 
-.. Create the `api` DNS record in a new public zone:
+.. Create the `api` DNS record in a new public zone.
```
Comment (codyhoag) [inline modules/installation-creating-azure-dns.adoc:99]: Should we change the period back to a colon? A colon is used in the next `api-int` DNS record creation command.

File: modules/installation-azure-limits.adoc [hunk 4]
```diff
@@ -160,16 +160,19 @@ If your applications create more Kubernetes `LoadBalancer` service objects,
...
 |Each cluster creates the following
 link:https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-overview[load balancers]:
 
 [horizontal]
+//`default`:: Public IP address that load balances requests to ports 80 and 443 across worker machines
+//`api`:: Public IP address that load balances requests to ports 6443 and 22623 across control plane machines
 `default`:: Public IP address that load balances requests to ports 80 and 443 across worker machines
-`api`:: Public IP address that load balances requests to ports 6443 and 22623 across control plane machines
+`internal`:: Private IP address that load balances requests to ports 6443 and 22623 across control plane machines
+`external`:: Public IP address that load balances requests to port 6443 across control plane machines
```
Comment (codyhoag) [inline modules/installation-azure-limits.adoc:175]: I think lines 147-179 can be consolidated into this:
```
|Network load balancers
| 3
ifndef::ash[]
| 1000 per region
endif::ash[]
|Each cluster creates the following
link:https://docs.microsoft.com/en-us/azure/load-balancer/load-balancer-overview[load balancers]:

[horizontal]
`default`:: Public IP address that load balances requests to ports 80 and 443 across worker machines
`internal`:: Private IP address that load balances requests to ports 6443 and 22623 across control plane machines
`external`:: Public IP address that load balances requests to port 6443 across control plane machines

If your applications create more Kubernetes `LoadBalancer` service objects,
your cluster uses more load balancers.
```
The changes that required the additional ifdefs are no longer present with these updates, from what I can tell.

**Review by codyhoag — COMMENTED** [review]
Just a few comments, including a suggestion on the `ifndef` at line 148.

File: modules/installation-creating-azure-dns.adoc (hunk 1 above)
Comment (mjpytlak) [reply modules/installation-creating-azure-dns.adoc]: Agreed. Done.

File: modules/installation-creating-azure-dns.adoc (hunk 2 above)
Comment (mjpytlak) [reply modules/installation-creating-azure-dns.adoc]: Thanks for catching these.

File: modules/installation-creating-azure-dns.adoc (hunk 3 above)
Comment (mjpytlak) [reply modules/installation-creating-azure-dns.adoc]: Yes. Oversight on my part.

File: modules/installation-azure-limits.adoc (hunk 4 above)
Comment (mjpytlak) [reply modules/installation-azure-limits.adoc:158]: I thought so. Thank you for verifying.

File: modules/installation-creating-azure-dns.adoc [hunk 5]
```diff
@@ -19,6 +19,13 @@ You must configure networking and load balancing in Microsoft {cp} for your
 {product-title} cluster to use. One way to create these components is
 to modify the provided Azure Resource Manager (ARM) template.
 
+ifdef::ash[]
+Load balancing requires the following DNS records:
+
+* An `api` DNS record in the public zone for the API public load balancer.
+* An `api-int` DNS record in the private zone for the API internal load balancer.
+endif::ash[]
```
Comment (patrickdillon) [inline modules/installation-creating-azure-dns.adoc:27]:
```suggestion
* An `api` DNS record for the API public load balancer in the DNS zone .
* An `api-int` DNS record for the API internal load balancer in the DNS zone.
endif::ash[]
```
The main change I am recommending here is to remove "private DNS zone." Azure Stack does not have "private" zones. It does not actually have "public" zones either, but just "DNS zones."

File: modules/installation-creating-azure-dns.adoc [hunk 6]
```diff
@@ -94,19 +112,19 @@ $ az network dns record-set a add-record -g ${BASE_DOMAIN_RESOURCE_GROUP} -z ${B
 ----
 
 ifdef::ash[]
-.. Create the `api-int` DNS record in a new public zone:
+.. Create the `api-int` DNS record in a new private zone:
```
Comment (patrickdillon) [inline modules/installation-creating-azure-dns.adoc:115]:
```suggestion
.. Create the `api-int` DNS record in a new DNS zone:
```

File: modules/installation-creating-azure-dns.adoc [hunk 7]
```diff
@@ -94,19 +112,19 @@ $ az network dns record-set a add-record -g ${BASE_DOMAIN_RESOURCE_GROUP} -z ${B
...
 +
 [source,terminal]
 ----
-$ az network dns record-set a add-record -g ${BASE_DOMAIN_RESOURCE_GROUP} -z "${CLUSTER_NAME}.${BASE_DOMAIN}" -n api-int -a ${PUBLIC_IP} --ttl 60
+$ az network dns record-set a add-record -g ${BASE_DOMAIN_RESOURCE_GROUP} -z "${CLUSTER_NAME}.${BASE_DOMAIN}" -n api-int -a ${PRIVATE_IP} --ttl 60
 ----
 +
-If you are adding the cluster to an existing public zone, you can create the `api-int` DNS
+If you are adding the cluster to an existing private zone, you can create the `api-int` DNS
```
Comment (patrickdillon) [inline modules/installation-creating-azure-dns.adoc:122]:
```suggestion
If you are adding the cluster to an existing DNS zone, you can create the `api-int` DNS
```

**Review by patrickdillon — COMMENTED** [review]
This looks good except Azure Stack does not have a "private" DNS zone. They do not actually have "public" DNS zones either, but just "DNS zones." So it is probably best to change previous usage of "public zones" to be consistent.

**Comment by patrickdillon** [comment]
nice work. LGTM

**Comment by tsze-redhat** [comment]
I reviewed the doc.
I haven't validated the final version of commands due to environment being down.
Otherwise,
LGTM

**Review by codyhoag — APPROVED** [review]
LGTM!

**Comment by codyhoag** [comment]
QE was able to verify and confirmed via Slack:
>To Hung Sze:

>I managed to create a cluster using the instruction here:
>https://deploy-preview-36950--osdocs.netlify.app/openshift-enterprise/latest/installing/installin[…]k_hub/installing-azure-stack-hub-user-infra.html
>I think it is the right one.
>If yes, I can mark those 3 BZ as verified.
>Thanks.

**Pull request merged**